### PR TITLE
Add way to upgrade from old forwarder

### PR DESCRIPTION
**Pull request by democrazyfi**
If you setup the grave on an older forwarder we need to provide a way to upgrade to the latest one

### CODE REVIEW

**Comment by tehnalogos** [comment]
@democrazyfi we need to unsubscribe from old forwarder and then we need to also do step 3 of the subscription flow when upgrading. otherwise new forwarder wont take effect for incoming assets and the old forwarder address will still have perms

File: frontend/constants/networks.ts [hunk 1]
```diff
@@ -22,6 +24,10 @@ const NETWORKS = {
     rpcUrl: 'https://rpc.testnet.lukso.network',
     explorerURL: 'https://explorer.execution.testnet.lukso.network',
     universalGraveForwarder: '0x7e77c704be7a6bba042f66eaba0b9557a872902d',
+    previousGraveForwarders: [
+      '0x1e2f66d2dC19bB13617D1d06CBb3E8d8aa4567e3',
+      '0x9c27a05310dC8aF53B60124B244cc9d12f202cdF',
```
Comment (democrazyfi) [inline frontend/constants/networks.ts:29]: you can test this change by putting this value into line 26 `universalGraveForwarder` and removing it from here, subscribe then revert the change to simulate an upgrade

File: frontend/utils/urdUtils.ts [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+  const urd8 = getChecksumAddress(URDLsp8)!;
+  let urd7Index = networkConfig.previousGraveForwarders.indexOf(urd7);
+  let urd8Index = networkConfig.previousGraveForwarders.indexOf(urd8);
+  if (urd7Index > -1 && urd8Index > -1 && urd7 === urd8) {
+    return networkConfig.previousGraveForwarders[urd7Index];
+  }
+  return null;
+};
+
+/**
```
Comment (democrazyfi) [inline frontend/utils/urdUtils.ts:43]: The functions below are copied from JoinGraveBtn 
I'll raise another PR to de-duplicate them

File: frontend/components/UpgradeURD.tsx [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+    completeText: { text: '', address: null },
+    complete: false,
+  },
+  {
+    title: 'Direct all 🆙 spam to the GRAVE',
+    completeText: { text: 'SPAM IS DEAD', address: null },
+    complete: false,
+  },
+  {
+    title: 'Migrate your GRAVE to the new forwarder',
```
Comment (tehnalogos) [inline frontend/components/UpgradeURD.tsx:51]: migration step needs to happen before setting new forwarder as LSP delegate. the reason is to avoid a broken state for the full flow. if the user executes steps 1-3, but not 4 they're in a broken state. meaning if they get an asset sent to their UP the transaction will revert because no grave vault exists associated to new forwarder. we want to direct spam to GRAVE only once we've set up everything up to work. @democrazyfi

File: frontend/components/UpgradeURD.tsx [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+        isClosable: true,
+      });
+      setLeavingStep(0);
+      setIsSubmitting(false);
+    }
+
+    setLeavingStep(3);
+
+    try {
+      const existingGrave = await oldForwarder.connect(signer).getGrave();
```
Comment (tehnalogos) [inline frontend/components/UpgradeURD.tsx:157]: given that you've set up the rest of the step as imports from urdUtils, this should be similarly abstracted in such a way. @democrazyfi

File: frontend/components/UpgradeURD.tsx (hunk 3 above)
Comment (democrazyfi) [reply frontend/components/UpgradeURD.tsx:51]: Yeah I had them the other way but the reason why I moved them is because when we setUrd7/8 the view changes state to the subscribed view

But I'll move  those bits to the end of the flow so that doesn't happen

File: frontend/components/UpgradeURD.tsx (hunk 4 above)
Comment (democrazyfi) [reply frontend/components/UpgradeURD.tsx:157]: I thought we'd do that just for things that need to be shared but perhaps for consistencey we can do that

File: frontend/components/UpgradeURD.tsx [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+  },
+];
+
+export const UpgradeURD = ({
+  oldForwarderAddress,
+}: {
+  oldForwarderAddress: string;
+}) => {
+  const walletContext = useContext(WalletContext);
+  const bgColor = useColorModeValue('light.green.brand', 'dark.purple.200');
```
Comment (Ledespana) [inline frontend/components/UpgradeURD.tsx:63]: do you need useColorModeValue if we 'technically' use always the same theme?

File: frontend/components/UpgradeURD.tsx [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+      index: -1,
+      count: leaveSteps.length,
+    });
+
+  const toast = useToast();
+
+  const migrateGraveToNewForwarder = async () => {
+    const oldForwarder = new ethers.Contract(
+      oldForwarderAddress,
+      LSP1GraveForwarderAbi.abi,
```
Comment (Ledespana) [inline frontend/components/UpgradeURD.tsx:84]: if it is an old forwarder, the abi will not be the same

File: frontend/constants/networks.ts [hunk 7]
```diff
@@ -25,6 +27,10 @@ const NETWORKS = {
     rpcUrl: 'https://rpc.testnet.lukso.network',
     explorerURL: 'https://explorer.execution.testnet.lukso.network',
     universalGraveForwarder: '0x7e77c704be7a6bba042f66eaba0b9557a872902d',
+    previousGraveForwarders: [
+      '0x1e2f66d2dC19bB13617D1d06CBb3E8d8aa4567e3',
```
Comment (Ledespana) [inline frontend/constants/networks.ts:31]: you may have to categorize them to find the right ABI to use
(ex: v1: 0x43wedfasd, V2: 0x0430239)

File: frontend/components/UpgradeURD.tsx (hunk 5 above)
Comment (democrazyfi) [reply frontend/components/UpgradeURD.tsx:63]: As discussed we'll keep this and remove it later in one go

File: frontend/components/UpgradeURD.tsx (hunk 6 above)
Comment (democrazyfi) [reply frontend/components/UpgradeURD.tsx:84]: We can store multiple ABIs once we make a change to getGrave

File: frontend/constants/networks.ts (hunk 7 above)
Comment (democrazyfi) [reply frontend/constants/networks.ts:31]: As long as the getGrave() function remains the same we should be okay
